In [9]:
# Anil kumar is very good customer, bought 3045 sales orders from our portal. We got 2000 profit because of anil kumar
customer_name = input("Enter the name of customer:")
number_of_sales_orders =  input("enter sales orders:")
profit = input("enter profit:")
s =  customer_name+" is very good customer, bought "+number_of_sales_orders
print(s)

Enter the name of customer: tej
enter sales orders: 1245
enter profit: 123547


tej is very good customer, bought 1245


In [13]:
# Anil kumar is very good customer, bought 3045 sales orders from our portal. We got 2000 profit because of anil kumar
customer_name = input("Enter the name of customer:")
number_of_sales_orders =  input("enter sales orders:")
profit = input("enter profit:")
s =  customer_name+" is very good customer, bought "+number_of_sales_orders+" sales orders from our portal. We got "+profit+" profit because of "+customer_name 
print(s)

Enter the name of customer: tej
enter sales orders: 12465
enter profit: 235632


tej is very good customer, bought 12465 sales orders from our portal. We got 235632 profit because of tej


### \ (Backslash) in the string "..." + "sales orders..." is used to split a long string across multiple lines for better readability. It doesn't affect the string itself.

In [11]:
# Anil kumar is very good customer, bought 3045 sales orders from our portal. We got 2000 profit because of anil kumar
customer_name = input("Enter the name of customer:")
number_of_sales_orders =  input("enter sales orders:")
profit = input("enter profit:")
s =  customer_name+" is very good customer, bought "+number_of_sales_orders+\
" sales orders from our portal. We got "+profit+" profit because of "+customer_name 
print(s)

Enter the name of customer: tej
enter sales orders: 145623
enter profit: 12346


tej is very good customer, bought 145623sales orders from our portal. We got 12346 profit because of tej


### Formatted String (f-string):
* The f-string allows you to embed the value of a variable (in this case, customer_name) directly into a string by placing the variable inside curly braces {}.
* The f before the string indicates that this is a formatted string.

In [1]:
# Anil kumar is very good customer, bought 3045 sales orders from our portal. We got 2000 profit because of anil kumar
customer_name = input("Enter the name of customer:")
print("customer_name is very good customer, bought")

Enter the name of customer: tej


customer_name is very good customer, bought


In [3]:
# Anil kumar is very good customer, bought 3045 sales orders from our portal. We got 2000 profit because of anil kumar
customer_name = input("Enter the name of customer:")
print("{customer_name} is very good customer, bought")

Enter the name of customer: tej


{customer_name} is very good customer, bought


In [5]:
# Anil kumar is very good customer, bought 3045 sales orders from our portal. We got 2000 profit because of anil kumar
customer_name = input("Enter the name of customer:")
print(f"{customer_name} is very good customer, bought") #formatted string

Enter the name of customer: tej


tej is very good customer, bought


In [12]:
customer_name = input("Enter the name of customer:")
number_of_sales_orders =  input("enter sales orders:")
profit = input("enter profit:")
print(f"{customer_name} is very good customer, bought {number_of_sales_orders} sales orders from our portal. We got {profit} profit because of {customer_name}")

Enter the name of customer: tej
enter sales orders: 10000
enter profit: 230043


tej is very good customer, bought 10000 sales orders from our portal. We got 230043 profit because of tej


In [14]:
a = 1000
print("The product cost:"+a)

TypeError: can only concatenate str (not "int") to str

### a reference to an integer object. 'The product cost' is a string, and we cannot concatenate a string with an integer, which is why a TypeError occurs.

In [16]:
a = 1000
print("The product cost:"+ str(a))

The product cost:1000


In [18]:
a = 1000
print(f"The product cost:{a}")

The product cost:1000


In [20]:
a = 12.456789
print(f"The product cost:{a}")

The product cost:12.456789


### Formatting Specifier:
* **: f:** This is a formatting specifier. It tells Python how to display the cost value.
* The space before the **f** indicates that the formatted number should include a leading space if it's positive (or a minus sign if it's negative).
* **f** means to format the number as a **floating-point number**.
* We get upto 6 decimal places by default.

In [24]:
cost = 12.456789
print(f"The product cost:{cost: f}")

The product cost: 12.456790


In [26]:
cost = 12.4567897852
print(f"The product cost:{cost: f}")

The product cost: 12.456790


* **.2** specifies that you want 2 decimal places. The number will be rounded to 2 decimal places.

In [30]:
cost = 12.4567897852
print(f"The product cost:{cost: .2f}")

The product cost: 12.46


### Format specifiers:
* The format specifiers %s, %f, %d, and %r are used in different ways to format strings, floating-point numbers, integers, and representations of objects in Python. Here’s what each one means:

#### 1. %s – String Formatting
* **Purpose:** This is used to insert (or convert) a string into the placeholder.
* **Usage:** Can accept any data type, but it will convert the value to a string.
#### 2. %f – Floating-Point Number Formatting
* **Purpose:** This is used to format floating-point numbers (numbers with decimals).
* **Usage:** By default, it shows 6 decimal places, but you can control the number of decimal places using %.nf where n is the number of decimals.
#### 3. %d – Integer Formatting
* **Purpose:** This is used to format integers (whole numbers). It converts the value to an integer if necessary (like rounding down floats).
* **Usage:** If you use %d with a float, the float will be converted to an integer by discarding the decimal part.
#### 4. %r – Representation (Repr) Formatting
* **Purpose:** This is used to insert the “official” string representation of an object, typically the one returned by repr(). It is mostly used for debugging and provides a more unambiguous string format, especially useful with complex data types.
* **Usage:** You can use %r for any Python object, and it will show the internal representation of the object.
| Format Specifier | Description                               | Example                                 |
|------------------|-------------------------------------------|-----------------------------------------|
| `%s`             | String (or converts anything to string)   | `"Hello, %s" % "world"` → `Hello, world`|
| `%f`             | Floating-point number (default 6 decimals)| `"%.2f" % 12.45678` → `12.46`           |
| `%d`             | Integer                                   | `"%d" % 25.9` → `25`                    |
| `%r`             | Representation of an object (debugging)   | `"%r" % [1, 2, 3]` → `[1, 2, 3]`        |


In [28]:
name = "Alice"
print("Hello, %s!" % name)


Hello, Alice!


In [30]:
cost = 12.456789
print("The product cost: %.2f" % cost)


The product cost: 12.46


In [35]:
cost = 12.9
print("Cost: %d" % cost)


Cost: 12


In [37]:
data = [1, 2, 3]
print("Data: %r" % data)


Data: [1, 2, 3]


### Special Symbols
* `\n` is used to print the statement on the next line.
* `\t` is used for tab spacing in the statement.
* To tell Python not to treat special symbols as such, we use a backslash `\`.
* If we place an **r** before the quotes ("") in a print statement, the string is treated as a **raw string**.

In [2]:
print("fsfga \n dfghjkl \")

fsfga 
 dfghjkl 	 dfghjk


In [4]:
print("fsfga\ndfghjkl\tdfghjk")

fsfga
dfghjkl	dfghjk


In [11]:
print("fsfga\\ndfghjkl\\tdfghjk")

fsfga\ndfghjkl\tdfghjk


In [17]:
print(len("\n"))

1


In [15]:
print(len("\\n"))

2


In [19]:
print(r"fsfga\ndfghjkl\tdfghjk")

fsfga\ndfghjkl\tdfghjk
